### ISSUE TITLE
Losing model accuracy by using cudnn.TemporalConvolution 

### ISSUE THREAD

**Issue by shuzi** [soumith/cudnn.torch#161]
Hi
https://github.com/shuzi/sentclassifier/blob/master/train.lua#L11_L17

I tested cunn.TemporalConvolution
             fbcunn.TemporalConvolutionFB
             cudnn.TemporalConvolution

TemporalConvolution from cunn and fbcunn give same model accuracy
TemporalConvolution of cudnn will give different model accuracy (few percent lower)
Any idea why?
I am using cudnn v4. Cuda 7.5

**Comment by ngimel** [comment]
Most probably the problem is cudnn's non-deterministic algorithm used for updateGradInput. The summations are performed in the different order in cudnn.TemporalConvolution and nn.TemporalConvolution, and it is quite possible the cudnn's error is larger.

**Comment by shuzi** [comment]
hmm,  the difference is too much, I lose more than 5% accuracy......

**Comment by ngimel** [comment]
In the script below, for a variety of sizes I get relative errors between nn.TemporalConvolution and cudnn.TemporalConvolution on the order of 1e-6. Please let me know if you get bigger errors for your parameters. 
```
require 'cudnn'
require 'cunn'
bs = 128
ki = 3
si = 1
ini = 6
print("feature Size, absolute error, relative error")
for iF=128, 1024,32 do 
	oF = iF
	input = torch.randn(bs,ini,iF):cuda()
	conv = cudnn.TemporalConvolution(iF,oF,ki):cuda()
	convg = nn.TemporalConvolution(iF,oF,ki):cuda()
	convg.weight = conv.weight:clone()
	convg.bias = conv.bias:clone()
	conv:zeroGradParameters()
	convg:zeroGradParameters()
	o1 = conv:updateOutput(input)
	o2 = convg:updateOutput(input)
	local diff =  (o1-o2):abs():max()
	print("forward diff", iF, diff, diff/o1:abs():max())
	local gO = torch.Tensor():resize(o1:size()):cuda()
	gO:rand(gO:nElement()):resize(o1:size())
	b1 = conv:backward(input, gO)
	b2 = convg:backward(input, gO)
	diff =  (b1-b2):abs():max()
	print("data diff", iF, diff, diff/b1:abs():max())
	diff =  (conv.gradWeight-convg.gradWeight):abs():max()
	print("weight diff", iF, diff, diff/conv.gradWeight:abs():max())
	diff =  (conv.gradBias-convg.gradBias):abs():max()
	print("bias diff", iF, diff, diff/conv.gradBias:abs():max())
end
```

**Comment by shuzi** [comment]
@ngimel Thank you!
Here is the result.
```
minwei@digits01:~/sentclassifier$ th test.lua
feature Size, absolute error, relative error	
forward diff	128	0	0	
data diff	128	1.4901161193848e-08	7.4708325622812e-08	
weight diff	128	4.9591064453125e-05	9.1132435477885e-07	
bias diff	128	0.000274658203125	9.9675898279195e-07	
forward diff	160	0	0	
data diff	160	1.4901161193848e-08	6.8916648208491e-08	
weight diff	160	4.9591064453125e-05	9.2413229309211e-07	
bias diff	160	0.000244140625	8.9567953302852e-07	
forward diff	192	0	0	
data diff	192	1.4901161193848e-08	9.9263632594326e-08	
weight diff	192	4.9591064453125e-05	8.492389251405e-07	
bias diff	192	0.0002593994140625	9.4018374066149e-07	
forward diff	224	2.6822090148926e-07	8.8230951770805e-07	
data diff	224	1.4901161193848e-08	9.0862263802409e-08	
weight diff	224	6.103515625e-05	1.0180604561204e-06	
bias diff	224	0.0002593994140625	9.5889672276175e-07	
forward diff	256	3.5762786865234e-07	1.2231889718913e-06	
data diff	256	1.4901161193848e-08	1.0598852101922e-07	
weight diff	256	4.9591064453125e-05	8.8443610836955e-07	
bias diff	256	0.000244140625	8.9450318845662e-07	
forward diff	288	2.9057264328003e-07	9.9242843267762e-07	
data diff	288	1.4901161193848e-08	1.0652230763536e-07	
weight diff	288	6.4849853515625e-05	1.0511068557084e-06	
bias diff	288	0.0001983642578125	7.1143590531642e-07	
forward diff	320	2.9802322387695e-07	1.1953804522423e-06	
data diff	320	1.4901161193848e-08	1.0415946230386e-07	
weight diff	320	6.4849853515625e-05	1.0628179154089e-06	
bias diff	320	0.0002288818359375	8.3623078007175e-07	
forward diff	352	3.8743019104004e-07	1.463721329077e-06	
data diff	352	1.4901161193848e-08	1.1876179898473e-07	
weight diff	352	7.62939453125e-05	1.4323969515728e-06	
bias diff	352	0.000244140625	8.8905049851284e-07	
forward diff	384	3.129243850708e-07	1.4248600075043e-06	
data diff	384	7.4505805969238e-09	6.3964106413758e-08	
weight diff	384	6.103515625e-05	9.2875917889794e-07	
bias diff	384	0.0002593994140625	9.4560670014634e-07	
forward diff	416	3.2782554626465e-07	1.4090248683282e-06	
data diff	416	7.4505805969238e-09	6.4957960507119e-08	
weight diff	416	5.7220458984375e-05	1.025221263253e-06	
bias diff	416	0.00030517578125	1.108882503697e-06	
forward diff	448	2.9802322387695e-07	1.2449431963543e-06	
data diff	448	7.4505805969238e-09	6.3269547506319e-08	
weight diff	448	6.103515625e-05	1.0859499523675e-06	
bias diff	448	0.000274658203125	9.9722284517872e-07	
forward diff	480	1.3411045074463e-07	6.4994413369091e-07	
data diff	480	7.4505805969238e-09	6.4426596845416e-08	
weight diff	480	6.103515625e-05	1.0354511285835e-06	
bias diff	480	0.0002288818359375	8.1253672666005e-07	
forward diff	512	2.9802322387695e-08	1.5113001042268e-07	
data diff	512	7.4505805969238e-09	7.5339214814703e-08	
weight diff	512	6.866455078125e-05	1.1090885990498e-06	
bias diff	512	0.000244140625	8.6495837387826e-07	
forward diff	544	3.427267074585e-07	1.734080033222e-06	
data diff	544	7.4505805969238e-09	7.4157855971799e-08	
weight diff	544	6.866455078125e-05	1.0942716883281e-06	
bias diff	544	0.000244140625	8.8454420432263e-07	
forward diff	576	3.129243850708e-07	1.3778100024938e-06	
data diff	576	7.4505805969238e-09	7.4514571100835e-08	
weight diff	576	6.866455078125e-05	1.0839344000492e-06	
bias diff	576	0.000274658203125	9.979364891042e-07	
forward diff	608	3.0547380447388e-07	1.5799087398471e-06	
data diff	608	7.4505805969238e-09	7.4748963082384e-08	
weight diff	608	6.103515625e-05	1.0370777327355e-06	
bias diff	608	0.000244140625	8.7571003664409e-07	
forward diff	640	4.0233135223389e-07	2.160394401602e-06	
data diff	640	7.4505805969238e-09	7.421023060755e-08	
weight diff	640	6.4849853515625e-05	1.0064194166485e-06	
bias diff	640	0.000244140625	8.9503219095154e-07	
forward diff	672	3.7252902984619e-07	2.01031161197e-06	
data diff	672	7.4505805969238e-09	7.5358097913379e-08	
weight diff	672	5.5313110351562e-05	8.3369699771337e-07	
bias diff	672	0.000244140625	8.8626338285404e-07	
forward diff	704	3.2782554626465e-07	1.7402422638717e-06	
data diff	704	7.4505805969238e-09	7.5147898579194e-08	
weight diff	704	7.2479248046875e-05	1.1427313564883e-06	
bias diff	704	0.000274658203125	9.9666869027324e-07	
forward diff	736	2.9802322387695e-07	1.641996760012e-06	
data diff	736	7.4505805969238e-09	8.3587019938848e-08	
weight diff	736	6.4849853515625e-05	1.0940273379412e-06	
bias diff	736	0.000274658203125	1.0006813528144e-06	
forward diff	768	3.129243850708e-07	1.9066666690878e-06	
data diff	768	7.4505805969238e-09	7.5739535977188e-08	
weight diff	768	7.62939453125e-05	1.2939111388411e-06	
bias diff	768	0.000274658203125	9.9093230875337e-07	
forward diff	800	4.3213367462158e-07	2.6490562280454e-06	
data diff	800	7.4505805969238e-09	8.5928658075348e-08	
weight diff	800	7.2479248046875e-05	1.1872485259216e-06	
bias diff	800	0.000274658203125	9.9345654328562e-07	
forward diff	832	3.5762786865234e-07	2.236691128612e-06	
data diff	832	7.4505805969238e-09	7.3566368046065e-08	
weight diff	832	6.4849853515625e-05	9.6760140377335e-07	
bias diff	832	0.000274658203125	1.0093974906378e-06	
forward diff	864	3.427267074585e-07	2.0701222407183e-06	
data diff	864	7.4505805969238e-09	9.9052621110649e-08	
weight diff	864	6.4849853515625e-05	1.0683840584278e-06	
bias diff	864	0.000274658203125	1.0070022460626e-06	
forward diff	896	3.427267074585e-07	2.2104691470886e-06	
data diff	896	7.4505805969238e-09	8.0091188623719e-08	
weight diff	896	6.6757202148438e-05	1.06608203946e-06	
bias diff	896	0.000274658203125	1.0013803471763e-06	
forward diff	928	3.5017728805542e-07	2.2923224171964e-06	
data diff	928	7.4505805969238e-09	8.8130248046108e-08	
weight diff	928	6.4849853515625e-05	9.4278336490918e-07	
bias diff	928	0.000274658203125	9.9488518472254e-07	
forward diff	960	3.5017728805542e-07	2.3614922018001e-06	
data diff	960	7.4505805969238e-09	8.3609104028119e-08	
weight diff	960	6.866455078125e-05	1.1085929627258e-06	
bias diff	960	0.0003204345703125	1.1542890964313e-06	
forward diff	992	4.0978193283081e-07	2.7558586799677e-06	
data diff	992	7.4505805969238e-09	8.4023248224505e-08	
weight diff	992	6.866455078125e-05	1.0545770577698e-06	
bias diff	992	0.00030517578125	1.104015968487e-06	
forward diff	1024	3.9488077163696e-07	2.5440082222346e-06	
data diff	1024	7.4505805969238e-09	9.061911796969e-08	
weight diff	1024	6.103515625e-05	9.8761524312526e-07	
bias diff	1024	0.000274658203125	9.9143641290374e-07	
```

**Comment by ngimel** [comment]
So in your output all relative errors are small. Are those TemporalConvolution sizes representative of what you actually are trying to run? I still don't see any reason other than fp accuracy for the accuracy difference in your training results.

**Comment by shuzi** [comment]
@ngimel 

Yes, the sizes are representative.
The classification task I am running is a difficult one: 7000 training examples and 3600 labels.
I guess this task is very sensitive to the backpropagation errors even the order is only 1e-7.
